### ISSUE TITLE
CC005 causes CC006 combinations to balloon

### ISSUE THREAD

**Issue by valentine195** [apigee/apigeelint#483]
It appears that CC005 is overwriting the constant value of the AST for the `TruthTable` [here](https://github.com/apigee/apigeelint/blob/main/lib/package/plugins/CC005-unterminatedStrings.js#L63).

That appears to cause the number of combinations being considered in CC006 to balloon due to the string processing of the node vs boolean [here](https://github.com/apigee/apigeelint/blob/4e64b928a8bc5111af132cd94bea0580a98857fc/lib/package/TruthTable.js#L781).

Anything over ~5 conditions evaluating to true ends up hanging for us.

I verified that disabling CC005 prevents this from occurring and fixed it by initializing a new variable for the value in CC005's callback.

**Comment by DinoChiesa** [comment]
hi @valentine195 
I understand what you wrote, and thanks for your diagnosis. 

Can you explain what this means? 
> Anything over ~5 conditions evaluating to true ends up hanging for us.

Or better, can you just provide an example for me, that exhibits the problem?  Our test suite doesn't trigger this problem and I'm not clear on what WOULD trigger this problem.

**Comment by valentine195** [comment]
Sure, reading it again I realize it is a little confusing 😅 

A policy in a proxy in our enterprise with this `FaultRule` causes ~15k+ combinations:

```xml
<FaultRules>
        <FaultRule name="serviceCalloutError">
            <Step>
                <Name>AssignServiceCalloutErrorResponse</Name>
            </Step>
            <Step>
                <Name>InitializeResponseLog</Name>
            </Step>
            <Step>
                <Name>FlowCalloutLogging</Name>
            </Step>
            <Condition>servicecallout.GetAutoPolicyServiceCallout.failed == true ||
                        servicecallout.GetAllAutoVehiclesServiceCallout.failed == true || 
                        servicecallout.GetAllAutoCoveragesServiceCallout.failed == true || 
                        servicecallout.GetAllAutoDriversServiceCallout.failed == true || 
                        servicecallout.SubmitAutoPolicyChangeServiceCallout.failed == true ||
                        servicecallout.GetAnAutoVehicleServiceCallout.failed == true
            </Condition>
        </FaultRule>
    </FaultRules>
```

Unfortunately this proxy has several policies similar to this (and some with even more), all of which cause thousands of combinations for the `TruthTable` to evaluate.

**Comment by DinoChiesa** [comment]
ok got it. Thanks for that. 
I'll get that fixed.

**Comment by DinoChiesa** [comment]
I am trying that Condition expression in my test API proxy.  I don't see a cascade of errors in CC006, or a hang.  Can you provide a more thorough explanation of the scenario that is leading to the failure you see?  Maybe a full API proxy (ideally reduced to minimum required to show the failure).   I want to see the failure first-hand here, before I introduce a change to fix it .

**Comment by valentine195** [comment]
Sure!

This is what I can see happening. This line rewrites the `boolean` `true` to be a `string` `"true"` inside the `TruthTable`'s AST:

https://github.com/apigee/apigeelint/blob/4e64b928a8bc5111af132cd94bea0580a98857fc/lib/package/plugins/CC005-unterminatedStrings.js#L63

Because of that, when the `getCombinationsFromVars` function is called, it processes each AST node as a `string` instead of a `boolean` [here](https://github.com/apigee/apigeelint/blob/4e64b928a8bc5111af132cd94bea0580a98857fc/lib/package/TruthTable.js#L781). This causes the possible values array to have 5 elements instead of 2, and when the cartesian product is taken, we're getting ~15k combinations instead of the 64 this combo _should_ be.

Like I said, this proxy actually has _multiple_ `FaultRules` with a similar condition set up (some have even more), so it causes a big slow down. Copying the value in CC005 instead of mutating fixes the issue for me.

**Comment by DinoChiesa** [comment]
yep - I got that, I got your diagnosis from your original message .  and I understand what you're saying.  

What I DO NOT SEE is any error during my tests.  When I drop that FaultRules section that you provided into my proxy, the plugins work just fine.   Maybe you could share a complete API proxy, instead of just a FaultRules stanza?   

I just want to reproduce the problem here.

**Comment by valentine195** [comment]
Hey, I actually did some more digging here. While the diagnosis above still stands for the example I gave, the condition actually does return after it. The scanning of this proxy actually hangs for this fault rule:

```xml
  <FaultRules>
    <FaultRule name="Generic Fault">
      <Step>
        <Name>AssignMessageGenericFault</Name>
      </Step>
      <Condition>(servicecallout.AutoPolicyChangeServiceCallout.failed == true || 
        servicecallout.GetAutoPolicyChangeByIdServiceCallout.failed == true || 
        servicecallout.DeleteAutoPolicyChangeServiceCallout.failed == true || 
        servicecallout.RateAutoPolicyChangeServiceCallout.failed == true || 
        servicecallout.UpdateAutoPolicyChangeServiceCallout.failed == true ||
        servicecallout.GetAdditionalInterestPolicyChangeServiceCallout.failed == true || 
        servicecallout.AdditionalInterestPolicyChangeServiceCallout.failed == true || 
        servicecallout.GetAdditionalInterestByIdServiceCallout.failed == true || 
        servicecallout.UpdateAdditionalInterestByIdServiceCallout.failed == true || 
        servicecallout.DeleteAdditionalInterestByIdServiceCallout.failed == true)
    </Condition>
    </FaultRule>
    <FaultRule name="DestinationUnreachableFault">
      <Step>
        <Name>AssignMessageDestinationUnreachableFault</Name>
      </Step>
    </FaultRule>
  </FaultRules>
```

This hangs in the `cartesianProduct` function because of the `String` mutation above, it generates hundreds of thousands of combinations. Here is a small proxy bundle that recreates the issue for me:

[apiproxy.zip](https://github.com/user-attachments/files/17600609/apiproxy.zip)

**Comment by ssvaidyanathan** [comment]
Fixed and released in v2.58.0
Can you please test?

**Comment by valentine195** [comment]
Yes, looks good now! Thank you!
